engine/collection/compliance/core.go

```diff
@@ -261,32 +282,57 @@ func (c *Core) processBlockProposal(proposal *messages.ClusterBlockProposal) err
 		Logger()
 	log.Info().Msg("processing block proposal")
 
+	// retrieve the parent block
+	//  - once we reach this point, the parent block must have been validated
+	//    and inserted to the protocol state
+	parent, err := c.headers.ByBlockID(header.ParentID)
+	if err != nil {
+		return fmt.Errorf("could not retrieve proposal parent: %w", err)
+	}
+
+	hotstuffProposal := model.ProposalFromFlow(header, parent.View)
+	err = c.validator.ValidateProposal(hotstuffProposal)
+	if err != nil {
+		if model.IsInvalidBlockError(err) {
+			return engine.NewInvalidInputErrorf("invalid block proposal: %w", err)
+		}
+		if errors.Is(err, model.ErrViewForUnknownEpoch) {
+			// We have received a proposal, but we don't know the epoch its view is within.
+			// We know:
+			//  - the parent of this block is valid and inserted (ie. we knew the epoch for it)
+			//  - if we then see this for the child, one of two things must have happened:
+			//    1. the proposer malicious created the block for a view very far in the future (it's invalid)
+			//      -> in this case we can disregard the block
+			//    2. no blocks have been finalized the epoch commitment deadline, and the epoch end
+			//       (breaking a critical assumption - see EpochCommitSafetyThreshold in protocol.Params for details)
+			//      -> in this case, the network has encountered a critical failure
+			//  - we assume in general that Case 2 will not happen, therefore we can discard this proposal
+			return engine.NewUnverifiableInputError("cannot validate proposal with view from unknown epoch: %w", err)
+		}
+		return fmt.Errorf("unexpected error validating proposal: %w", err)
+	}
+
 	// see if the block is a valid extension of the protocol state
 	block := &cluster.Block{
 		Header:  proposal.Header,
 		Payload: proposal.Payload,
 	}
-	err := c.state.Extend(block)
-	// if the block proposes an invalid extension of the protocol state, then the block is invalid
-	if state.IsInvalidExtensionError(err) {
-		return engine.NewInvalidInputErrorf("invalid extension of protocol state (block: %x, height: %d): %w",
-			header.ID(), header.Height, err)
-	}
-	// protocol state aborted processing of block as it is on an abandoned fork: block is outdated
-	if state.IsOutdatedExtensionError(err) {
-		return engine.NewOutdatedInputErrorf("outdated extension of protocol state: %w", err)
-	}
+	err = c.state.Extend(block)
```

Comment:
Can we extend godoc of `Extend` with returned errors? Will be easier to check if all sentinels are covered. Thanks

Reply:
https://github.com/onflow/flow-go/pull/3294/commits/08a04b3b9a70d2791af3dc9148afed7af71b5aa9